### ISSUE TITLE
Ambiguous use of a named function

### ISSUE THREAD

**Issue by thereforeiam** [google/closure-compiler#1247]
The following code : 
```javascript
var a;

if(!a)
{
	a = function a () {};
	function b() {};
}

a();
b();
```
Issue the following error when compiled : 
```
test.js:10: ERROR - Ambiguous use of a named function: b.
b();
^

1 error(s), 0 warning(s)
```

Is it normal ?
Following ES5 scoping rules, I find nothing "ambiguous" about this code.

**Comment by MatrixFrog** [comment]
In ES6, functions are block scoped, so b is not declared where it is called. However, in ES5 a function declaration in a block is in scope even outside the block, though the details differ between browsers. The ES6 spec has an [annex](http://www.ecma-international.org/ecma-262/6.0/#sec-block-level-function-declarations-web-legacy-compatibility-semantics) which to be honest gets a bit confusing.

To simplify things, we warn about some cases that are technically valid in ES5, so that users can just keep in mind the semantics of ES6 excluding that annex. I believe this warning is suppressible but I would recommend you rearrange your code (for example by changing `function b() {}` to `var b = function() {}`).

**Comment by thereforeiam** [comment]
I agree with you about rearranging the code.
However the compiler does not warn, it exits with an error, even with  --language_in ECMASCRIPT5
That does not seems correct to me.
I would be fine with a warning

**Comment by blickly** [comment]
The ES5 spec only allows function declarations inside "SourceElements", not inside blocks.

**Comment by thereforeiam** [comment]
@blickly the compiler does not issue error for
```javascript
if(...)
{
	function c() { }
}
```
as of today

**Comment by nphollon** [comment]
@thereforeiam Even though it is an error, you can still suppress it with either of the following:

    --jscomp_warning ambiguousFunctionDecl
    --jscomp_off ambiguousFunctionDecl
